packages/stratimikos/adapters/belos/tpetra/Thyra_BelosTpetraPreconditionerFactory_def.hpp

```diff
@@ -0,0 +1,369 @@
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(thyraTpetraFwdOp));
+
+  typedef Tpetra::Operator<scalar_type, local_ordinal_type, global_ordinal_type, node_type> TpetraLinOp;
+  const Teuchos::RCP<const TpetraLinOp> tpetraFwdOp = thyraTpetraFwdOp->getConstTpetraOperator();
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(tpetraFwdOp));
+
+  // Belos-specific typedefs:
+  typedef Tpetra::MultiVector<scalar_type, local_ordinal_type, global_ordinal_type, node_type>     TpetraMV; 
+  typedef Belos::TpetraOperator<scalar_type, local_ordinal_type, global_ordinal_type, node_type> BelosTpOp;
+  typedef Belos::LinearProblem<scalar_type, TpetraMV, TpetraLinOp> BelosTpLinProb;
+
+#ifdef THYRA_BELOS_PREC_ENABLE_HALF_PRECISION
+  // CAG: There is nothing special about the combination double-float,
+  //      except that I feel somewhat confident that Trilinos builds
+  //      with both scalar types.
+  typedef typename Teuchos::ScalarTraits<scalar_type>::halfPrecision half_scalar_type;
+  typedef Tpetra::Operator<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type> TpetraLinOpHalf;
+  typedef Tpetra::MultiVector<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type>     TpetraMVHalf; 
+  typedef Belos::TpetraOperator<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type> BelosTpOpHalf;
+  typedef Belos::LinearProblem<half_scalar_type, TpetraMVHalf, TpetraLinOpHalf> BelosTpLinProbHalf;
+#endif
+
+  // Retrieve concrete preconditioner object
+
+  const Teuchos::Ptr<DefaultPreconditioner<scalar_type> > defaultPrec =
+    Teuchos::ptr(dynamic_cast<DefaultPreconditioner<scalar_type> *>(prec));
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(defaultPrec));
+
+  // This check needed to address Issue #535. 
+  Teuchos::RCP<const Teuchos::ParameterList> innerParamList;
+  if (paramList_.is_null ()) {
+    innerParamList = Teuchos::rcp(new Teuchos::ParameterList(*getValidParameters()));
+  }
+  else {
+    innerParamList = paramList_;
+  }
+
+  bool useHalfPrecision = false;
+  if (innerParamList->isParameter("half precision"))
+    useHalfPrecision = innerParamList->get<bool>("half precision");
+
+  const std::string solverType = Teuchos::getParameter<std::string>(*innerParamList, "BelosPrec Solver Type");
+  const Teuchos::RCP<Teuchos::ParameterList> packageParamList = 
+          Teuchos::rcp(new Teuchos::ParameterList(*Teuchos::sublist(innerParamList, "BelosPrec Solver Params")));
+
+  // solverTypeUpper is the upper-case version of solverType.
+  std::string solverTypeUpper (solverType);
+  if (solverTypeUpper.size () > 0) {
+    for (size_t k = 0; k < solverTypeUpper.size (); ++k) {
+      solverTypeUpper[k] = ::toupper(solverTypeUpper[k]);
+    }
+  }
+
+  // Create the initial preconditioner
+  if (Teuchos::nonnull(out) && Teuchos::includesVerbLevel(verbLevel, Teuchos::VERB_LOW)) {
+    *out << "\nCreating a new BelosTpetra::Preconditioner object...\n";
```

Comment:
Do you really want to print this at `Teuchos::VERB_LOW`?  Don't remember how this plays with other printing in the stack of objects.